### PR TITLE
feat: Use lookup tables in ts_lex

### PR DESCRIPTION
**Pull request by osiewicz**
Summary: This pull request augments `tree-sitter-cli` with ability to generate lookup tables instead of if-else chains. This should reduce compile times of `tree-sitter-rust` and improve binary sizes all around. 

Detailed summary [is available here](https://github.com/osiewicz/ramblings/blob/tree-sitter-lut/posts/tree-sitter/001-lookup-tables/0001-lookup-tables.md) - it's also WIP.

### CODE REVIEW

**Comment by clason** [comment]
A good test case would be tree-sitter-julia, which is one of the girthiest parsers I know of.

**Comment by amaanq** [comment]
Lean, Verilog, & C# are also very large

**Comment by osiewicz** [comment]
Gotcha, I will make sure to include these in size comparisons then. I've ran them mostly across languages included in `test/fixtures/grammars`.
Thanks =)

**Comment by amaanq** [comment]
This might be of use to you

https://github.com/sogaiu/ts-questions

**Review by vicspina — APPROVED** [review]

**Comment by clason** [comment]
Come on, don't approve PRs if you don't have merge privileges or your review was not explicitly requested. Don't be that guy!

**Comment by clason** [comment]
<details><summary>Here's the full list of parsers supported by nvim-treesitter, sorted descending by size:</summary>

```
 18M Jul 20 12:54 verilog.so
8.9M Jul 20 12:54 julia.so
7.9M Jul 20 12:53 d.so
6.7M Jul 20 12:54 c_sharp.so
5.4M Jul 20 12:54 perl.so
5.3M Jul 20 12:54 objc.so
4.9M Jul 20 12:54 ocaml.so
4.8M Jul 20 12:54 tlaplus.so
4.4M Jul 20 12:54 ocaml_interface.so
4.2M Jul 20 12:54 haskell.so
3.9M Jul 20 12:54 hlsl.so
3.6M Jul 20 12:54 kotlin.so
3.4M Jul 20 12:53 cuda.so
3.4M Jul 20 12:54 scala.so
3.0M Jul 20 12:53 arduino.so
2.6M Jul 20 12:53 swift.so
2.4M Jul 20 12:54 latex.so
2.3M Jul 20 12:53 cpp.so
2.1M Jul 20 12:53 ruby.so
2.0M Jul 20 12:53 agda.so
1.8M Jul 20 12:53 hack.so
1.7M Jul 20 12:53 elixir.so
1.7M Jul 20 12:53 v.so
1.5M Jul 20 12:53 llvm.so
1.4M Jul 20 12:53 mlir.so
1.4M Jul 20 12:53 odin.so
1.3M Jul 20 12:53 qmljs.so
1.2M Jul 20 12:53 vim.so
1.2M Jul 20 12:53 hoon.so
1.2M Jul 20 12:53 gitcommit.so
1.2M Jul 20 12:53 tsx.so
1.1M Jul 20 12:53 typescript.so
969K Jul 20 12:53 dart.so
952K Jul 20 12:53 sql.so
889K Jul 20 12:53 fortran.so
872K Jul 20 12:53 zig.so
792K Jul 20 12:53 commonlisp.so
775K Jul 20 12:53 ispc.so
744K Jul 20 12:53 rust.so
728K Jul 20 12:53 php.so
678K Jul 20 12:54 pascal.so
663K Jul 20 12:53 pony.so
630K Jul 20 12:53 glsl.so
581K Jul 20 12:53 vala.so
568K Jul 20 12:53 org.so
535K Jul 20 12:53 bash.so
485K Jul 20 12:53 slint.so
469K Jul 20 12:53 c.so
453K Jul 20 12:53 r.so
422K Jul 20 12:53 markdown.so
421K Jul 20 12:53 python.so
421K Jul 20 12:53 markdown_inline.so
420K Jul 20 12:53 solidity.so
404K Jul 20 12:53 java.so
404K Jul 20 12:53 m68k.so
389K Jul 20 12:53 squirrel.so
357K Jul 20 12:53 awk.so
357K Jul 20 12:53 cairo.so
356K Jul 20 12:54 smali.so
340K Jul 20 12:53 javascript.so
324K Jul 20 12:53 poe_filter.so
323K Jul 20 12:53 ql.so
309K Jul 20 12:53 pug.so
309K Jul 20 12:53 starlark.so
307K Jul 20 12:53 ada.so
292K Jul 20 12:53 dhall.so
276K Jul 20 12:55 wing.so
275K Jul 20 12:53 sparql.so
259K Jul 20 12:53 gleam.so
259K Jul 20 12:53 go.so
259K Jul 20 12:53 erlang.so
259K Jul 20 12:53 systemtap.so
244K Jul 20 12:53 cue.so
231K Jul 20 12:53 terraform.so
231K Jul 20 12:53 hcl.so
227K Jul 20 12:53 mermaid.so
227K Jul 20 12:53 yang.so
214K Jul 20 12:53 yaml.so
213K Jul 20 12:53 elm.so
212K Jul 20 12:53 gdscript.so
211K Jul 20 12:53 phpdoc.so
211K Jul 20 12:53 hurl.so
210K Jul 20 12:53 fsh.so
210K Jul 20 12:53 prql.so
195K Jul 20 12:53 t32.so
195K Jul 20 12:53 ledger.so
179K Jul 20 12:53 matlab.so
179K Jul 20 12:53 scss.so
164K Jul 20 12:53 nickel.so
163K Jul 20 12:54 supercollider.so
163K Jul 20 12:53 elvish.so
162K Jul 20 12:53 vimdoc.so
162K Jul 20 12:53 make.so
148K Jul 20 12:53 haskell_persistent.so
147K Jul 20 12:53 racket.so
146K Jul 20 12:53 puppet.so
146K Jul 20 12:53 robot.so
146K Jul 20 12:53 hare.so
131K Jul 20 12:53 lalrpop.so
130K Jul 20 12:53 scheme.so
130K Jul 20 12:53 rego.so
130K Jul 20 12:53 bicep.so
130K Jul 20 12:53 thrift.so
130K Jul 20 12:53 luadoc.so
130K Jul 20 12:53 meson.so
130K Jul 20 12:53 twig.so
116K Jul 20 12:53 beancount.so
114K Jul 20 12:53 kdl.so
114K Jul 20 12:53 teal.so
114K Jul 20 12:53 hocon.so
114K Jul 20 12:53 fennel.so
114K Jul 20 12:53 clojure.so
114K Jul 20 12:53 jq.so
 99K Jul 20 12:53 cooklang.so
 99K Jul 20 12:53 fish.so
 98K Jul 20 12:53 jsonnet.so
 98K Jul 20 12:53 firrtl.so
 98K Jul 20 12:53 luau.so
 98K Jul 20 12:53 wgsl_bevy.so
 98K Jul 20 12:53 tablegen.so
 98K Jul 20 12:53 cmake.so
 98K Jul 20 12:53 nix.so
 98K Jul 20 12:53 usd.so
 98K Jul 20 12:53 func.so
 98K Jul 20 12:53 proto.so
 98K Jul 20 12:53 htmldjango.so
 86K Jul 20 12:53 rst.so
 82K Jul 20 12:53 ron.so
 82K Jul 20 12:54 wgsl.so
 81K Jul 20 12:53 capnp.so
 81K Jul 20 12:53 blueprint.so
 73K Jul 20 12:53 vue.so
 73K Jul 20 12:53 astro.so
 70K Jul 20 12:53 svelte.so
 66K Jul 20 12:53 lua.so
 66K Jul 20 12:53 foam.so
 66K Jul 20 12:53 uxntal.so
 66K Jul 20 12:53 menhir.so
 66K Jul 20 12:53 css.so
 66K Jul 20 12:53 fusion.so
 66K Jul 20 12:53 tiger.so
 66K Jul 20 12:53 pioasm.so
 66K Jul 20 12:53 prisma.so
 65K Jul 20 12:53 dot.so
 65K Jul 20 12:53 devicetree.so
 65K Jul 20 12:53 rasi.so
 65K Jul 20 12:53 dockerfile.so
 65K Jul 20 12:53 smithy.so
 65K Jul 20 12:53 graphql.so
 51K Jul 20 12:53 html.so
 50K Jul 20 12:53 glimmer.so
 50K Jul 20 12:53 yuck.so
 49K Jul 20 12:53 gitattributes.so
 49K Jul 20 12:53 json5.so
 49K Jul 20 12:53 turtle.so
 49K Jul 20 12:53 query.so
 49K Jul 20 12:53 ninja.so
 49K Jul 20 12:53 surface.so
 49K Jul 20 12:53 heex.so
 49K Jul 20 12:53 promql.so
 34K Jul 20 12:53 ocamllex.so
 34K Jul 20 12:53 comment.so
 34K Jul 20 12:54 rnoweb.so
 34K Jul 20 12:53 janet_simple.so
 34K Jul 20 12:53 godot_resource.so
 34K Jul 20 12:53 toml.so
 33K Jul 20 12:53 bass.so
 33K Jul 20 12:53 luap.so
 33K Jul 20 12:53 http.so
 33K Jul 20 12:53 chatito.so
 33K Jul 20 12:53 cpon.so
 33K Jul 20 12:53 diff.so
 33K Jul 20 12:53 ebnf.so
 33K Jul 20 12:53 elsa.so
 33K Jul 20 12:53 gosum.so
 33K Jul 20 12:53 ini.so
 33K Jul 20 12:53 json.so
 33K Jul 20 12:53 bibtex.so
 33K Jul 20 12:53 git_config.so
 33K Jul 20 12:53 gitignore.so
 33K Jul 20 12:53 hjson.so
 33K Jul 20 12:53 jsonc.so
 33K Jul 20 12:53 ungrammar.so
 33K Jul 20 12:53 embedded_template.so
 33K Jul 20 12:53 git_rebase.so
 33K Jul 20 12:53 corn.so
 33K Jul 20 12:53 groovy.so
 33K Jul 20 12:53 passwd.so
 33K Jul 20 12:53 pem.so
 33K Jul 20 12:53 qmldir.so
 33K Jul 20 12:53 regex.so
 33K Jul 20 12:53 vhs.so
 33K Jul 20 12:53 po.so
 33K Jul 20 12:53 gomod.so
 33K Jul 20 12:53 gowork.so
 33K Jul 20 12:53 jsdoc.so
 33K Jul 20 12:53 sxhkdrc.so
 33K Jul 20 12:53 todotxt.so
 33K Jul 20 12:53 eex.so
```
</details>

**Comment by amaanq** [comment]
I wonder if my latest changes to the Rust parser improved compilation times @osiewicz, mind trying out 0.20.4?

**Comment by osiewicz** [comment]
Yeah, it looks like they've improved a bit;
```
tree-sitter-rust/src (0431a2c) [?] via C v14.0.3-clang
❯ hyperfine ' cc -O3  -c -fPIC -arch arm64 -I ../src -Wall -Wextra -Wno-unused-parameter -Wno-unused-but-set-variable -Wno-trigraphs parser.c'
Benchmark 1:  cc -O3  -c -fPIC -arch arm64 -I ../src -Wall -Wextra -Wno-unused-parameter -Wno-unused-but-set-variable -Wno-trigraphs parser.c
  Time (mean ± σ):     29.787 s ±  0.509 s    [User: 29.689 s, System: 0.084 s]
  Range (min … max):   28.940 s … 30.426 s    10 runs


tree-sitter-rust/src (0431a2c) [?] via C v14.0.3-clang took 4m58s
❯ git checkout v0.20.4
Previous HEAD position was 0431a2c v0.20.3
HEAD is now at afb6000 docs: update README

tree-sitter-rust/src (afb6000) [?] via C v14.0.3-clang
❯ hyperfine ' cc -O3  -c -fPIC -arch arm64 -I ../src -Wall -Wextra -Wno-unused-parameter -Wno-unused-but-set-variable -Wno-trigraphs parser.c'
Benchmark 1:  cc -O3  -c -fPIC -arch arm64 -I ../src -Wall -Wextra -Wno-unused-parameter -Wno-unused-but-set-variable -Wno-trigraphs parser.c
  Time (mean ± σ):     28.931 s ±  1.148 s    [User: 28.835 s, System: 0.083 s]
  Range (min … max):   27.431 s … 30.325 s    10 runs

```

**Comment by osiewicz** [comment]
I've noticed that `parser.c` is quick to compile again (~10x quicker) with clang 15.0 (bundled with Xcode 15). I guess then that I'll close this PR as it doesn't seem like the extra complexity is worth it; runtime/binary size wins were rather small.
Tl;dr: if you find that tree-sitter grammars take a while to compile, make sure you're not on clang 14.
